packages/seacas/libraries/ioss/src/Ioss_ScopeGuard.h

```diff
@@ -0,0 +1,297 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef SCOPEGUARD_H_
+#define SCOPEGUARD_H_
+
+/*
+  Scopeguard, by Andrei Alexandrescu and Petru Marginean, December 2000.
+  Modified by Joshua Lehrer, FactSet Research Systems, November 2005.
+*/
+
+template <class T> class RefHolder
+{
+  T &ref_;
+
+public:
+  explicit RefHolder(T &ref) : ref_(ref) {}
+  operator T &() const { return ref_; }
+
+private:
+  // Disable assignment - not implemented
+  RefHolder &operator=(const RefHolder &);
+};
+
+template <class T> inline RefHolder<T> ByRef(T &t) { return RefHolder<T>(t); }
+
+class ScopeGuardImplBase
+{
+  ScopeGuardImplBase &operator=(const ScopeGuardImplBase &);
+
+protected:
+  ~ScopeGuardImplBase() = default;
+  ScopeGuardImplBase(const ScopeGuardImplBase &other) throw() : dismissed_(other.dismissed_)
+  {
+    other.Dismiss();
+  }
+  template <typename J> static void SafeExecute(J &j) throw()
+  {
+    if (!j.dismissed_) {
+      try {
+        j.Execute();
+      }
+      catch (...) {
+      }
+    }
+  }
+
+  mutable bool dismissed_;
+
+public:
+  ScopeGuardImplBase() throw() : dismissed_(false) {}
+  void Dismiss() const throw() { dismissed_ = true; }
+};
+
+// typedef const ScopeGuardImplBase& ScopeGuard;
+__attribute__((unused)) typedef const ScopeGuardImplBase &ScopeGuard;
+
+template <typename F> class ScopeGuardImpl0 : public ScopeGuardImplBase
+{
+public:
+  static ScopeGuardImpl0<F> MakeGuard(F fun) { return ScopeGuardImpl0<F>(fun); }
+  ~ScopeGuardImpl0() throw() { SafeExecute(*this); }
+  void Execute() { fun_(); }
+
+protected:
+  explicit ScopeGuardImpl0(F fun) : fun_(fun) {}
+  F fun_;
+};
+
+template <typename F> inline ScopeGuardImpl0<F> MakeGuard(F fun)
+{
+  return ScopeGuardImpl0<F>::MakeGuard(fun);
+}
+
+template <typename F, typename P1> class ScopeGuardImpl1 : public ScopeGuardImplBase
+{
+public:
+  static ScopeGuardImpl1<F, P1> MakeGuard(F fun, P1 p1) { return ScopeGuardImpl1<F, P1>(fun, p1); }
+  ~ScopeGuardImpl1() throw() { SafeExecute(*this); }
+  void Execute() { fun_(p1_); }
+
+protected:
+  ScopeGuardImpl1(F fun, P1 p1) : fun_(fun), p1_(p1) {}
+  F        fun_;
+  const P1 p1_;
+};
+
+template <typename F, typename P1> inline ScopeGuardImpl1<F, P1> MakeGuard(F fun, P1 p1)
+{
+  return ScopeGuardImpl1<F, P1>::MakeGuard(fun, p1);
+}
+
+template <typename F, typename P1, typename P2> class ScopeGuardImpl2 : public ScopeGuardImplBase
```

Comment:
C++11 has variadic templates, so you don't need to have separate ScopeGuardImpl0, ScopeGuardImpl1, ScopeGuardImpl2, etc. classes.